pep-0584.rst

```diff
@@ -1,1043 +1,974 @@
-                raise KeyError('duplicate key')
-            return super().__add__(other)
+=========
+Rationale
+=========
 
-3.  **Add the values (as Counter does).**  Too specialised to be used as
-    the default behaviour.
+The new operators will have the same relationship to the
+``dict.update`` method as the list concatenate (``+``) and extend
+(``+=``) operators have to ``list.extend``.
 
-4.  **First seen wins: value-preserving semantics.**  It isn't clear that
-    this behaviour has many use-cases.  Probably too specialised to be used
-    as the default, best to leave this for subclasses, or simply reverse
-    the order of the arguments::
+Key conflicts will be resolved by keeping the rightmost value.  This
+matches the existing behavior of similar ``dict`` operations, where
+the last seen value always wins::
 
-        # d1 merged with d2, keeping existing values in d1
-        d2 + d1
+    {'a': 1, 'a': 2}
+    {**d, **e}
+    d.update(e)
+    d[k] = v
+    {k: v for x in (d, e) for (k, v) in x.items()}
 
-5.  **Concatenate values in a list**::
+All of the above follow the same rule.  This PEP takes the position
+that this behavior is simple, obvious, usually the behavior we want,
+and should be the default behavior for dicts.  This means that dict
+union is not commutative; in general ``d | e != e | d``.
 
-        {'a': 1} + {'a': 2} == {'a': [1, 2]}
+Similarly, the *iteration order* of the key-value pairs in the
+dictionary will follow the same semantics as the examples above, with
+each newly merged pair being appended (or moved) to the end of the
+current sequence.
 
-    This is likely to be too specialised to be the default.  It is not clear
-    what to do if the values are already lists::
 
-        {'a': [1, 2]} + {'a': [3, 4]}
+=============
+Specification
+=============
 
-    Should this give ``{'a': [1, 2, 3, 4]}`` or ``{'a': [[1, 2], [3, 4]]}``?
+Dict union will return a new dict containing the left operand merged
+with the right operand. If a key appears in both operands, the
+last-seen value (i.e. that from the right-hand operand) wins::
 
-To summarise this section: this PEP proposes option 1, **last seen wins** as
-the default behaviour, with alternatives left to subclasses of dict.
+    >>> d = {'spam': 1, 'eggs': 2, 'cheese': 3}
+    >>> e = {'cheese': 'cheddar', 'aardvark': 'Ethel'}
+    >>> d | e
+    {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
+    >>> e | d
+    {'aardvark': 'Ethel', 'spam': 1, 'eggs': 2, 'cheese': 3}
 
+The augmented assignment version operates in-place::
 
-Syntax
-------
+    >>> d |= e
+    >>> d
+    {'spam': 1, 'eggs': 2, 'cheese': 'cheddar', 'aardvark': 'Ethel'}
 
-An alternative to the ``+`` operator is the pipe ``|`` operator, which
-is used for set union.  This suggestion did not receive much support
-on Python-Ideas.
+Augmented assignment behaves identically to the ``update`` method, so
+it also accepts iterables of key-value pairs. This is analogous to
```

Comment:
What if the argument is a Mapping but not a dict? IIRC `dict.update()` gets this right. Of course the intended implementation (call `dict.update()`) means this operator will also get it right, but since you are specifying this you may have to specify the full behavior. (And a difference with `dict.update()` is that operators don't take keyword args of course. :-)

Reply:
I wasn't sure how specific to be here... let me know what you think of the new text:

> Augmented assignment behaves identically to the `update` method called with a single positional argument, so it also accepts anything implementing the Mapping protocol (more specifically, anything with the `keys` and `__getitem__` methods) or iterables of key-value pairs.